core/variable_model.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2017 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Components for the variable model.
+ * @author marisaleung@google.com (Marisa Leung)
+ */
+'use strict';
+
+goog.provide('Blockly.VariableModel');
+
+/**
+ * Class for a variable model.
+ * Holds information for the variable including name, id, and type. 
+ * @param {!string} name The name of the variable. This must be unique across
+ *     variables and procedures.
+ * @param {?string} opt_type The type of the variable like 'int' or 'string'.
+ *     Does not need to be unique. Field_variable can filter variables based on
+ *     their type. This will default to '' which is a specific type.
+ * @param {?string} opt_id The unique id of the variable. This will default to 
+ *     a UUID.
+ * @see {Blockly.FieldVariable}   
+ * @constructor
+ */
+Blockly.VariableModel = function(name, opt_type, opt_id) {
+  /*
```

Comment:
Woops my bad, I copy pasta'd from the design doc. Updated design doc too. Thank you!